File: app/src/main/java/ru/voronezhtsev/weatherapp/view/MainActivity.java
package ru.voronezhtsev.weatherapp.view;

import android.Manifest;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.widget.TextView;

import com.arellomobile.mvp.MvpAppCompatActivity;
import com.arellomobile.mvp.presenter.InjectPresenter;
import com.arellomobile.mvp.presenter.ProvidePresenter;

import java.util.List;

import ru.voronezhtsev.weatherapp.App;
import ru.voronezhtsev.weatherapp.R;
import ru.voronezhtsev.weatherapp.di.WeatherComponent;
import ru.voronezhtsev.weatherapp.net.models.forecast.Forecast;


public class MainActivity extends MvpAppCompatActivity implements MainView {

    private static final int REQUEST_FINE_LOCATION = 1;
    @InjectPresenter
    MainPresenter mMainPresenter;
    private TextView mCurrentTemp;
    private RecyclerView mForecastRecycler;

    @ProvidePresenter
    MainPresenter providePresenter() {
        WeatherComponent component = App.getInstance().getWeatherComponent();
        return new MainPresenter(component.getWeatherRepository(), component.getForecastsRepository(),
                component.getLocationRepository());
    }
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        mCurrentTemp = findViewById(R.id.current_temp);
        mForecastRecycler = findViewById(R.id.forecast_recycler_view);
        RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false);
        mForecastRecycler.setLayoutManager(layoutManager);
        requestPermissions(new String[]{Manifest.permission.ACCESS_FINE_LOCATION}, REQUEST_FINE_LOCATION);
    }

    @Override
    protected void onStart() {
        super.onStart();
    }

    @Override
    protected void onStop() {
        super.onStop();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
    }

    @Override
    public void showTemperature(double temp) {
        mCurrentTemp.setText(TempUtils.trimZeroes(Math.round(temp)));
    }

    @Override
    public void showForecast(List<Forecast> forecast) {
        mForecastRecycler.setAdapter(new ForecastAdapter(forecast));
    }
}

File: app/src/main/java/ru/voronezhtsev/weatherapp/presentation/MainActivity.kt
package ru.voronezhtsev.weatherapp.presentation

import android.Manifest
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.View.GONE
import androidx.recyclerview.widget.LinearLayoutManager
import com.arellomobile.mvp.MvpAppCompatActivity
import com.arellomobile.mvp.presenter.InjectPresenter
import com.arellomobile.mvp.presenter.ProvidePresenter
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.fragment_no_place.*
import ru.voronezhtsev.weatherapp.App.Companion.component
import ru.voronezhtsev.weatherapp.R
import ru.voronezhtsev.weatherapp.models.presentation.CityModel
import ru.voronezhtsev.weatherapp.models.presentation.WeatherModel

class MainActivity : MvpAppCompatActivity(), MainView, OnPlaceClickListener {

    companion object {
        private const val REQUEST_COARSE_LOCATION: Int = 1
    }
    @InjectPresenter
    internal lateinit var presenter: MainPresenter
    lateinit var cities: List<CityModel>;

    @ProvidePresenter
    fun providePresenter(): MainPresenter = component.mainPresenterFactory.get(0)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayShowTitleEnabled(false)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        requestPermissions(arrayOf(Manifest.permission.ACCESS_COARSE_LOCATION), REQUEST_COARSE_LOCATION)
        add_place_button.setOnClickListener { v: View -> presenter.addPlace() }
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.menu, menu)
        return true
    }

    override fun showNoPlacesAdded(cityList: List<CityModel>) {

        cities = cityList
        val fragmentTransaction = supportFragmentManager.beginTransaction()
        fragmentTransaction.add(container.id, NoPlaceFragment.newInstance("", ""));
        fragmentTransaction.addToBackStack(null);
        fragmentTransaction.commit()
    }

    override fun addPlace() {
        val fragmentTransaction = supportFragmentManager.beginTransaction()
        fragmentTransaction.replace(container.id, AddPlaceFragment.newInstance(cities))
        fragmentTransaction.addToBackStack(null)
        fragmentTransaction.commit()
    }

    override fun showWeather(weather: List<WeatherModel>) {
        noPlaceTextHolder.visibility = GONE
        weatherList.layoutManager = LinearLayoutManager(this)
        weatherList.adapter = WeatherAdapter(weather)
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        if (item?.itemId == android.R.id.home) {
            onBackPressed()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onClick(cityId: Long) {
        presenter.load(cityId)
    }
}